### ISSUE TITLE
jobpid() / b:terminal_job_id returning wrong PID

### ISSUE THREAD

**Issue by brettanomyces** [neovim/neovim#5478]
- `nvim --version`:
NVIM 0.1.6-dev
Build type: RelWithDebInfo
Compilation: /usr/bin/x86_64-linux-gnu-gcc -g -O2 -fstack-protector-strong -Wformat -Werror=format-secur
ity -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -DDISABLE_LOG -Wall -Wextra -pedantic -Wno
-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto
 -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -D_GNU_SOURCE -I/build/neovim-16z3RV/neovim-0.1.5ubunt
u1+git201610041714+2801+21~ubuntu16.04.1/build/config -I/build/neovim-16z3RV/neovim-.1.5ubuntu1+git201610041714+2801+21~ubuntu16.04.1/src -I/build/neovim-16z3RV/neovim-0.1.5ubuntu1+git201610041714+2801+21~ubuntu16.04.1/.deps/usr/include -I/build/neovim-16z3RV/neovim-0.1.5ubuntu1+git201610041714+2801+21~ubuntu16.04.1/.deps/usr/include -I/build/neovim-16z3RV/neovim-0.1.5ubuntu1+git201610041714+2801+21~ubuntu16.04.1/.deps/usr/include -I/build/neovim-16z3RV/neovim-0.1.5ubuntu1+git201610041714+2801+21~ubuntu16.04.1/.deps/usr/include -I/build/neovim-16z3RV/neovim-0.1.5ubuntu1+git201610041714+2801+21~ubuntu16.04.1/.deps/us
r/include -I/build/neovim-16z3RV/neovim-0.1.5ubuntu1+git201610041714+2801+21~ubuntu16.04.1/.deps/usr/include -I/usr/include -I/build/neovim-16z3RVneovim-.1.5ubuntu1+git201610041714+2801+21~ubuntu16.04.1/build/src/nvim/auto -I/build/neovim-16z3RV/neovim-0.1.5ubuntu1+git201610041714+2801+21~ubuntu16.04.1/build/include
Compiled by buildd@lcy01-23

Optional features included (+) or not (-): +acl   +iconv    +jemalloc +tui
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

- Vim (version: ) behaves differently?
- Operating system/version:  XUbuntu 16.04
- Terminal name/version: fishshell
- `$TERM`: xterm-256color

### Actual behaviour
```
nvim -u NONE
:terminal
ps
  PID TTY          TIME CMD
29048 pts/8    00:00:00 fish
29060 pts/8    00:00:00 ps
:echom jobpid(b:terminal_job_id)
29048
```
This is OK

```
nvim -u NONE
:terminal /usr/bin/fish
  PID TTY          TIME CMD
29071 pts/7    00:00:00 fish
29073 pts/7    00:00:00 fish
29085 pts/7    00:00:00 ps
:echom jobpid(b:terminal_job_id)
29071
```
This is not OK, the PID of the terminal buffer in Neovim is the second PID (29073). Not sure why when running `:terminal` without an argument no new shell is started but one is when providing an argument.

**Comment by jamessan** [comment]
As documented in `:help terminal`:

> Spawns {cmd} using the current value of 'shell' and
'shellcmdflag' in a new terminal buffer.  This is
equivalent to: >
                                                      
>      :enew
      :call termopen('{cmd}')
      :startinsert
                                                      
> If no {cmd} is given, 'shellcmdflag' will not be sent
to |termopen()|.

So, in the case of `:terminal`, the shell is directly spawned.  In the case of `:terminal /usr/bin/fish`, the shell (fish) is used to spawn the command (fish).

29071 _is_ the correct PID.  It's the parent shell (connected to nvim) that's running the child shell (29073).  You can see this if you run `ps afx`.

**Comment by brettanomyces** [comment]
Makes sense, however I think the help docs for the `b:terminal_job_pid` variable are a bit misleading.
```
- b:terminal_job_pid The PID of the top-level process running in the  terminal.
```

**Comment by jamessan** [comment]
How is that misleading?  The parent shell _is_ the top-level process running in the terminal.  It just happens that its sole use was to run another shell.

**Comment by brettanomyces** [comment]
It's a buffer local variable, for a terminal buffer, called terminal_job_pid that doesn't return the PID of job (the job to run a shell) running in the terminal buffer, except when it does (:terminal called with no args). 

I think the main problem is the "running in the terminal part", people (me) assume that means the terminal buffer, not the terminal that Neovim is running in.

**Comment by brettanomyces** [comment]
As a note to others that might have the same misunderstanding.

To get the PID of the shell running in the terminal buffer

- If the terminal buffer was starged using `:terminal` with no args:
  ```
  b:terminal_job_pid
  ```
- if the terminal buffer was open using `:terminal` with args or with `termopen()`
  ```
  substitute(strtrans(system("pgrep -P " . b:terminal_job_pid)), '\^@', '', 'g')
  ```

**Comment by justinmk** [comment]
We can add a way to get the immediate child pid.

**Comment by justinmk** [comment]
@brettanomyces You could also use `exec` command so that the child process replaces its parent.

    :te exec /usr/bin/fish

I wonder if it would be a good idea to do this by default for `:terminal {cmd}`...

**Comment by jamessan** [comment]
Or just execute the command directly instead of using the shell?  Having `{cmd}` default to the shell when `{cmd}` isn't given makes sense, but I'm not sure I understand when `{cmd}` is run through the shell.

**Comment by justinmk** [comment]
Oh, right :>

**Comment by justinmk** [comment]
What do most terminal emulators do here? Is it possible for example to run a process "directly" using xterm, or does xterm always wrap commands in a shell?

**Comment by jamessan** [comment]
`xterm -e ...` doesn't invoke an intermediate shell, iirc.  (I'm not at a computer I can test that with right now.)

**Comment by justinmk** [comment]
> In the case of :terminal /usr/bin/fish, the shell (fish) is used to spawn the command (fish).

This behavior is inherited from the difference between `jobstart({String})` vs `jobstart({List})`. The String variant always prepends `'shell'`, the List runs the argv directly.

We could try to tokenize the `:terminal` args, but I believe this has been trouble for us (and Vim). If that level of control is needed, `jobstart()` exists for that purpose.

> We can add a way to get the immediate child pid.

Turns out this is not easy. At least not that I can see. libuv doesn't have that info AFAICT, we don't track it, and the solutions I found are platform-specific hackery (e.g. querying `/proc/...`). Would really like to be proven wrong though :>
